File: lua/test/rolling_release_entity_test.lua
-- RollingRelease entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("vercel_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("RollingReleaseEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:RollingRelease(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = rolling_release_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "rolling_release." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set VERCEL_TEST_ROLLING_RELEASE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local rolling_release_ref01_ent = client:RollingRelease(nil)
    local rolling_release_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.rolling_release"), "rolling_release_ref01"))
    rolling_release_ref01_data["project_id"] = setup.idmap["project01"]

    local rolling_release_ref01_data_result, err = rolling_release_ref01_ent:create(rolling_release_ref01_data, nil)
    assert.is_nil(err)
    rolling_release_ref01_data = helpers.to_map(type(rolling_release_ref01_data_result) == 'table' and rolling_release_ref01_data_result.data_get and rolling_release_ref01_data_result:data_get() or rolling_release_ref01_data_result)
    assert.is_not_nil(rolling_release_ref01_data)

    -- UPDATE
    local rolling_release_ref01_data_up0_up = {
    }

    local rolling_release_ref01_markdef_up0_name = "advancementType"
    local rolling_release_ref01_markdef_up0_value = "Mark01-rolling_release_ref01_" .. tostring(setup.now)
    rolling_release_ref01_data_up0_up[rolling_release_ref01_markdef_up0_name] = rolling_release_ref01_markdef_up0_value

    local rolling_release_ref01_resdata_up0_result, err = rolling_release_ref01_ent:update(rolling_release_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local rolling_release_ref01_resdata_up0 = helpers.to_map(type(rolling_release_ref01_resdata_up0_result) == 'table' and rolling_release_ref01_resdata_up0_result.data_get and rolling_release_ref01_resdata_up0_result:data_get() or rolling_release_ref01_resdata_up0_result)
    assert.is_not_nil(rolling_release_ref01_resdata_up0)
    assert.are.equal(rolling_release_ref01_resdata_up0[rolling_release_ref01_markdef_up0_name], rolling_release_ref01_markdef_up0_value)

    -- LOAD
    local rolling_release_ref01_match_dt0 = {}
    local rolling_release_ref01_data_dt0_loaded, err = rolling_release_ref01_ent:load(rolling_release_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(rolling_release_ref01_data_dt0_loaded)


  end)
end)

function rolling_release_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/rolling_release/RollingReleaseTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read rolling_release test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "rolling_release01", "rolling_release02", "rolling_release03", "project01", "project02", "project03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("VERCEL_TEST_ROLLING_RELEASE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["VERCEL_TEST_ROLLING_RELEASE_ENTID"] = idmap,
    ["VERCEL_TEST_LIVE"] = "FALSE",
    ["VERCEL_TEST_EXPLAIN"] = "FALSE",
    ["VERCEL_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["VERCEL_TEST_ROLLING_RELEASE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["VERCEL_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["VERCEL_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["VERCEL_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["VERCEL_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/rolling_release_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/vercel-sdk/go"
	"github.com/voxgig-sdk/vercel-sdk/go/core"

	vs "github.com/voxgig-sdk/vercel-sdk/go/utility/struct"
)

func TestRollingReleaseEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.RollingRelease(nil)
		if ent == nil {
			t.Fatal("expected non-nil RollingReleaseEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := rolling_releaseBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "rolling_release." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set VERCEL_TEST_ROLLING_RELEASE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		rollingReleaseRef01Ent := client.RollingRelease(nil)
		rollingReleaseRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "rolling_release"}), "rolling_release_ref01"))
		rollingReleaseRef01Data["project_id"] = setup.idmap["project01"]

		rollingReleaseRef01DataResult, err := rollingReleaseRef01Ent.Create(rollingReleaseRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		rollingReleaseRef01Data = core.ToMapAny(entityData(rollingReleaseRef01DataResult))
		if rollingReleaseRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// UPDATE
		rollingReleaseRef01DataUp0Up := map[string]any{
		}

		rollingReleaseRef01MarkdefUp0Name := "advancementType"
		rollingReleaseRef01MarkdefUp0Value := fmt.Sprintf("Mark01-rolling_release_ref01_%d", setup.now)
		rollingReleaseRef01DataUp0Up[rollingReleaseRef01MarkdefUp0Name] = rollingReleaseRef01MarkdefUp0Value

		rollingReleaseRef01ResdataUp0Result, err := rollingReleaseRef01Ent.Update(rollingReleaseRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		rollingReleaseRef01ResdataUp0 := core.ToMapAny(entityData(rollingReleaseRef01ResdataUp0Result))
		if rollingReleaseRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if rollingReleaseRef01ResdataUp0[rollingReleaseRef01MarkdefUp0Name] != rollingReleaseRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", rollingReleaseRef01MarkdefUp0Name, rollingReleaseRef01ResdataUp0[rollingReleaseRef01MarkdefUp0Name])
		}

		// LOAD
		rollingReleaseRef01MatchDt0 := map[string]any{}
		rollingReleaseRef01DataDt0Loaded, err := rollingReleaseRef01Ent.Load(rollingReleaseRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if rollingReleaseRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}


	})
}

func rolling_releaseBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "rolling_release", "RollingReleaseTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read rolling_release test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse rolling_release test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"rolling_release01", "rolling_release02", "rolling_release03", "project01", "project02", "project03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("VERCEL_TEST_ROLLING_RELEASE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"VERCEL_TEST_ROLLING_RELEASE_ENTID": idmap,
		"VERCEL_TEST_LIVE":      "FALSE",
		"VERCEL_TEST_EXPLAIN":   "FALSE",
		"VERCEL_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["VERCEL_TEST_ROLLING_RELEASE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["VERCEL_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["VERCEL_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewVercelSDK(core.ToMapAny(mergedOpts))
	}

	live := env["VERCEL_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["VERCEL_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
